// A protocol definition for communication between a master controller
// and data collection sensors.
// This is an extension of the original protocol, intended to show how it can
// be updated without breaking devices that continue using the original.

// Fields added in version 2 are marked with the comment 'V2'. This is strictly
// for the humans; protobuf doesn't need such information.

syntax = "proto2";


// Contains information that identifies a particular sensor device.
message DeviceIdentification {
    required uint32     id              = 1;
    optional string     name            = 2;
    optional bool       keeps_history   = 3 [ default=false ]; // V2
}

// Sent by sensor devices to tell the controller about their existence.
message Connect {
    required DeviceIdentification dev_id = 1;
}

// Sent by the controller to give a command to a sensor device.
message Command {
    enum CommandType {
        REPORT_DATA             = 1; // Report latest data reading.
        REPORT_DATA_HISTORY     = 2; // V2; Report all data readings (for history-keeping devices)
        CLEAR_DATA_HISTORY      = 3; // V2; Clear all data readings (for history-keeping devices)
    }

    required CommandType cmd_type   = 1;
}

// Sent by sensor devices to report data.
message Report {
    required DeviceIdentification dev_id    = 1;

    // Note: Some values are scaled so that we can use the more-efficient
    // *int32 types instead of float/double.
    message SensorData {
        optional sint32 temperature             = 1; // Degrees Celsius * 10
        optional uint32 humidity                = 2; // Relative humidity * 10 
        optional uint32 rainfall                = 3; // In mm; V2
        optional uint32 pressure                = 4; // In mm Hg; V2
    }

    optional SensorData data                = 2;
    repeated SensorData data_history        = 3; // V2
}

// Top-level generic message wrapper used by both sensors and controllers.
message Msg {
    enum MsgType {
        CONNECT     = 0;
        COMMAND     = 1;
        REPORT      = 2;
    }

    // The 'type' field indicates which of the optional fields that follow it
    // should also be in this message. Without this discriminator, the receiver
    // would need to query for every optional field.
    required MsgType    msg_type    = 1;
    optional Connect    connect_msg = 2;
    optional Command    command_msg = 3;
    optional Report     report_msg  = 4;
}
